flogger/flogger-api-tests/src/test/kotlin/com/google/common/flogger/parser/PrintfMessageParserSpec.kt

```diff
@@ -0,0 +1,148 @@
+import io.kotest.matchers.string.shouldContain
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+
+internal class PrintfMessageParserSpec {
+
+    @Test
+    fun testPrintfNextTerm() {
+        nextPrintfTerm("", 0) shouldBe -1
+        nextPrintfTerm("%X", 0) shouldBe 0
+        nextPrintfTerm("Hello %X World %X", 0) shouldBe 6
+        nextPrintfTerm("Hello %X World %X", 6) shouldBe 6
+        nextPrintfTerm("Hello %X World %X", 7) shouldBe 15
+        nextPrintfTerm("Hello %% World %X", 0) shouldBe 15
+        nextPrintfTerm("Hello %X World %X", 16) shouldBe -1
+    }
+
+    @Test
+    fun testPrintfNextTermFails() {
+        val exception = assertThrows<ParseException> {
+            nextPrintfTerm("Hello %", 0)
+        }
+        exception.message shouldContain "[%]"
+    }
+
+    @Test
+    fun testParse() {
+        assertParse(FakeParser, "Hello World")
+        assertParse(FakeParser, "Hello %A %B %C World", "0:A", "1:B", "2:C")
+        assertParse(FakeParser, "Hello %1\$A %2\$B %3\$C World", "0:A", "1:B", "2:C")
+        assertParse(FakeParser, "Hello %2\$A %B %C %1\$D World", "1:A", "0:B", "1:C", "0:D")
+        assertParse(FakeParser, "Hello %A %<B %<C World", "0:A", "0:B", "0:C")
+        assertParse(FakeParser, "Hello %???X World", "0:???:X")
+        assertParse(FakeParser, "%%%A%%X%%%B%%", "0:A", "1:B")
+    }
+
+    @Test
+    fun testParsePrintfError() {
+        assertParseError(FakeParser, "%", "[%]")
+        assertParseError(FakeParser, "Hello %", "[%]")
+        // Unterminated parameter
+        assertParseError(FakeParser, "Hello %1", "[%1]")
+        assertParseError(FakeParser, "Hello %1$", "[%1$]")
+        // Missing index
+        assertParseError(FakeParser, "Hello %$ World", "[%$]")
+        // Leading zeros
+        assertParseError(FakeParser, "Hello %01\$X World", "[%01$]")
+        // Index too large
+        assertParseError(FakeParser, "Hello %1000000X World", "[%1000000]")
+        // Printf indices are 1-based
+        assertParseError(FakeParser, "Hello %0\$X World", "[%0$]")
+        // Relative indexing cannot come first.
+        assertParseError(FakeParser, "Hello %<X World", "[%<]")
+        // Unexpected flag or missing term character
+        assertParseError(FakeParser, "Hello %????", "[%????]")
+        assertParseError(FakeParser, "Hello %X %<", "[%<]")
+        // Gaps in term indices is a parse error, report the first argument index not referenced.
+        assertParseError(FakeParser, "Hello %X %100\$X World", "first missing index=1")
+    }
+
+    @Test
+    fun testGetSafeSystemNewline() {
+        // This should pass even if "line.separator" is set to something else.
+        val nl = PrintfMessageParser.getSafeSystemNewline()
+        nl shouldBeIn listOf("\n", "\r", "\r\n")
+    }
+
+    @Test
+    fun testUnescapePrintfSupportsNewline() {
+        val nl = PrintfMessageParser.getSafeSystemNewline()
+        assertThat(unescapePrintf("%n")).isEqualTo(nl)
```

Comment:
@yevhenii-nadtochii,  please migrate to Kotest assertions where it's possible. They read better.

Consider using Structural Replace in IDEA. It works for things like this like a charm.